pkg/fleetallocation/find_test.go

```diff
@@ -0,0 +1,48 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fleetallocation
+
+import (
+	"testing"
+
+	"agones.dev/agones/pkg/apis/stable/v1alpha1"
+	"github.com/stretchr/testify/assert"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestFindReadyGameServer(t *testing.T) {
```

Comment:
Add this test case: node1 with 4 allocated (first in list), then node 2 with 4 ready.

Node1 would be the trap I mentioned earlier.

Thing is that allocated is the primary sort key, but ready>0 is the veto condition. Actually you could also include this in the 'best' search above. 

Its quite late here so I am not sure, am I making sense? 😴

Reply:
I added the extra test - seems to be passing -- maybe I got it wrong? :man_shrugging: :smile:

Reply:
That's right, it was working before because you were setting `update=true` if bestGs=nil` as the first condition. If you were using 'bestCount=nil', it wouldn't have worked, and for someone who reads the code the reason is not easily seen as they should be an atomic pair. 

The rules (which you explained on Slack and in the design ticket) were very clear, but are not transposed into the code clearly imo. For example, the update logic is split into two places: first you set an `update` bool that is actually half of the truth and then you apply another condition. I will add a few notes inline for some specific things I would do differently.